src/model/enhance-images.ts

```diff
@@ -0,0 +1,333 @@
+import { JSDOM } from 'jsdom';
+
+/**
+ * decideCaption
+ * Decides if a text element should be used to caption the preceding images
+ * @param {TextBlockElement} element - The text element to examine
+ * @param {Boolean} isPhotoEssay - We treat photo essays differently
+ * @param {ImageBlockElement[]} imageBuffer - We use this to decide if a caption is needed or not
+ * */
+const decideCaption = ({
+	element,
+	isPhotoEssay,
+	imageBuffer,
+}: {
+	element?: TextBlockElement;
+	isPhotoEssay: boolean;
+	imageBuffer: ImageBlockElement[];
+}): string => {
+	//  the convention: <ul><li><Caption text</li></ul>
+	if (!element) return '';
+	// We only set a special caption on non photo essay articles if there are more than two
+	// halfWidth images in the buffer
+	const countOfHalfWidthImages = imageBuffer.filter(
+		(image) => image.role === 'halfWidth',
+	).length;
+	if (!isPhotoEssay && countOfHalfWidthImages < 2) return '';
+	// Extract the caption
+	const frag = JSDOM.fragment(element.html);
+	if (!frag || !frag.firstChild || !frag.firstElementChild) return '';
+	const hasULwrapper = frag.firstChild.nodeName === 'UL';
+	const containsLItags = frag.firstElementChild.outerHTML.includes('<li>');
```

Comment:
Belive it or not, [captions with multiple lines](https://www.theguardian.com/politics/2019/dec/10/on-the-road-with-boris-johnson-a-photo-essay?dcr=true) is a thing 🤷 . It's often used to separate the text out, one line for each image. I suspect actually this might be part of the reasoning for using ul/li as the 'convention' here?

![Screenshot 2021-04-12 at 13 34 25](https://user-images.githubusercontent.com/1336821/114395042-d4f11800-9b93-11eb-9183-75eb5ba0ce0b.jpg)

Reply:
oof, oh right!

Reply:
Hmm. When we do this properly in Composer, this is a great example: _Above_, _Right_, _Far right_ don’t work on mobile! 🤦‍♂️

Reply:
@paperboyo That's a great point! Examples like this are so much more powerful than developers complaining about apis